sycl/include/sycl/handler.hpp

```diff
@@ -939,6 +939,33 @@ class __SYCL_EXPORT handler {
     }
   }
 
+  void verifyDeviceHasProgressGuarantee(
+      sycl::ext::oneapi::experimental::forward_progress_guarantee guarantee,
+      sycl::ext::oneapi::experimental::execution_scope threadScope,
+      sycl::ext::oneapi::experimental::execution_scope coordinationScope);
+
+  template <
+      typename PropertyT,
+      sycl::ext::oneapi::experimental::forward_progress_guarantee Guarantee,
+      sycl::ext::oneapi::experimental::execution_scope CoordinationScope>
+  void extractForwardProgressPropArgsAndVerify(
```

Comment:
I'm not sure how having this not inlined is helping in readability. You have to duplicate all the `constexpr if` from the caller, I don't even see what common code this helper abstracts away.